File: package/main/src/Tool/rgbaToCmyk.ts
import { division } from "@/Math/division";
import { roundOf } from "@/Math/roundOf";
import { subtract } from "@/Math/subtract";

/**
 * rgbaをcmykに変換する
 * @param rgba
 * @returns { c: number, m: number, y: number, k: number, a: number }
 * @example rgbaToCmyk({ r: 0, g: 0, b: 0, a: 1 }); // { c: 0, m: 0, y: 0, k: 100, a: 1 }
 */
export const rgbaToCmyk = ({
  r,
  g,
  b,
  a = 1,
}: {
  r: number;
  g: number;
  b: number;
  a?: number;
}): { c: number; m: number; y: number; k: number; a: number } => {
  const rPrime = division(r, 255);
  const gPrime = division(g, 255);
  const bPrime = division(b, 255);

  const k = subtract(1, Math.max(rPrime, gPrime, bPrime));
  const c = division(subtract(subtract(1, rPrime), k), subtract(1, k)) || 0;
  const m = division(subtract(subtract(1, gPrime), k), subtract(1, k)) || 0;
  const y = division(subtract(subtract(1, bPrime), k), subtract(1, k)) || 0;

  return {
    c: roundOf(c * 100, 2),
    m: roundOf(m * 100, 2),
    y: roundOf(y * 100, 2),
    k: roundOf(k * 100, 2),
    a,
  };
};

File: package/umt_python/src/color/rgba_to_cmyk.py
from typing import Dict


def rgba_to_cmyk(r: int, g: int, b: int, a: float = 1.0) -> Dict[str, float]:
    """
    Convert RGBA color to CMYK color model.

    Args:
        r: Red value (0-255)
        g: Green value (0-255)
        b: Blue value (0-255)
        a: Alpha value (0-1)

    Returns:
        Dictionary with c, m, y, k as percentages 0-100 and alpha channel

    Raises:
        ValueError: If any input values are out of their valid ranges

    Example:
        >>> rgba_to_cmyk(0, 0, 0, 1)
        {'c': 0, 'm': 0, 'y': 0, 'k': 100, 'a': 1}
    """
    if r < 0 or r > 255 or g < 0 or g > 255 or b < 0 or b > 255 or a < 0 or a > 1:
        raise ValueError("Invalid rgba value")

    r_prime = r / 255
    g_prime = g / 255
    b_prime = b / 255

    k = 1 - max(r_prime, g_prime, b_prime)

    if k == 1:
        c = 0
        m = 0
        y = 0
    else:
        c = (1 - r_prime - k) / (1 - k)
        m = (1 - g_prime - k) / (1 - k)
        y = (1 - b_prime - k) / (1 - k)

    return {
        "c": round(c * 100, 2),
        "m": round(m * 100, 2),
        "y": round(y * 100, 2),
        "k": round(k * 100, 2),
        "a": a,
    }
